File: Python/binary_search.py
# searching an element X in a array(sorted) using Binary search
# Time Complexity = O(log n) | Space Complexity = O(log n)


def binarySearch(arr, target):
    start = 0
    end = len(arr) - 1
    while start <= end:
        #finding the middle element of arr
        mid = start + (end-start)//2
        if arr[mid] == target:
            return mid
        
        #target is present at left side of mid element
        if arr[mid] > target:
            end = mid - 1
        #target is present at right side of mid element
        else:
            start = mid + 1
    #if target is not present in arr
    return -1


# driver code
arr = [10, 20, 35, 60, 70, 110, 120, 140]
target = 35

result = binarySearch(arr,target)  # our defined function is stored in var result

if result == -1:
    print("Element doesn't exists !!")
else:
    print("Element is at index", result)

File: JavaScript(JS)/binarySearch.js

// arr is the array on which binary search is to be performed
// x is the search element
// start is the start index of the search
// end is the ending index of the search

let binarySearch = function (arr, x, start, end) {
      
    // Intialize start as the start point
    // end as the end of an array
    // Mid variable to be search in an array.

    let mid=Math.floor((start + end)/2);
    
    if (start > end) return false;
    else if (arr[mid]===x) return true;
    else if(arr[mid] > x)
        return binarySearch(arr, x, start, mid-1);
    else
         return binarySearch(arr, x, mid+1, end);
}
  
